File: apps/backend/src/makima/routes/marketplace.py
"""Marketplace API routes.

Provides REST endpoints for browsing, installing, and uninstalling MCP servers
from the marketplace.
"""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from structlog import get_logger

from makima.core.deps import get_current_user
from makima.auth.models import User
from makima.marketplace.models import (
    MarketplaceItem,
    MarketplaceItemListResponse,
    InstallRequest,
    InstallResponse,
    UninstallRequest,
    UninstallResponse,
    InstalledItemInfo,
)
from makima.marketplace.service import MarketplaceService

logger = get_logger(__name__)

router = APIRouter(prefix="/api/marketplace", tags=["marketplace"])

# Global service instance (initialized on app startup)
_marketplace_service: Optional[MarketplaceService] = None


def get_marketplace_service() -> MarketplaceService:
    """Get the marketplace service instance."""
    global _marketplace_service
    if _marketplace_service is None:
        _marketplace_service = MarketplaceService()
    return _marketplace_service


@router.get("/items", response_model=MarketplaceItemListResponse)
async def list_marketplace_items(
    search: Optional[str] = Query(None, description="Search query"),
    tags: Optional[list[str]] = Query(None, description="Filter by tags"),
    user: User = Depends(get_current_user),
) -> MarketplaceItemListResponse:
    """List all marketplace items with optional filtering.

    Args:
        search: Optional search query to filter items.
        tags: Optional list of tags to filter by.
        user: Current authenticated user.

    Returns:
        List of marketplace items with total count.
    """
    service = get_marketplace_service()

    try:
        items = service.get_items(search=search, tags=tags)
        return MarketplaceItemListResponse(items=items, total=len(items))
    except Exception as e:
        logger.error("Failed to list marketplace items", error=str(e))
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/items/{item_id}", response_model=MarketplaceItem)
async def get_marketplace_item(
    item_id: str,
    user: User = Depends(get_current_user),
) -> MarketplaceItem:
    """Get a single marketplace item by ID.

    Args:
        item_id: The unique item identifier.
        user: Current authenticated user.

    Returns:
        The marketplace item.

    Raises:
        HTTPException: 404 if item not found.
    """
    service = get_marketplace_service()

    item = service.get_item(item_id)
    if not item:
        raise HTTPException(status_code=404, detail=f"Marketplace item not found: {item_id}")

    return item


@router.get("/tags", response_model=list[str])
async def list_marketplace_tags(
    user: User = Depends(get_current_user),
) -> list[str]:
    """Get all unique tags from marketplace items.

    Args:
        user: Current authenticated user.

    Returns:
        Sorted list of unique tags.
    """
    service = get_marketplace_service()
    return service.get_tags()


@router.post("/install", response_model=InstallResponse)
async def install_marketplace_item(
    request: InstallRequest,
    user: User = Depends(get_current_user),
) -> InstallResponse:
    """Install an MCP server from the marketplace.

    Args:
        request: Installation request with item ID and parameters.
        user: Current authenticated user.

    Returns:
        Installation response with success status.

    Raises:
        HTTPException: 404 if item not found.
    """
    service = get_marketplace_service()

    item = service.get_item(request.item_id)
    if not item:
        raise HTTPException(status_code=404, detail=f"Marketplace item not found: {request.item_id}")

    try:
        response = service.install_item(
            item=item,
            target=request.target,
            selected_method_index=request.selected_method_index,
            parameters=request.parameters,
        )
        return response
    except Exception as e:
        logger.error("Failed to install marketplace item", item_id=request.item_id, error=str(e))
        return InstallResponse(
            success=False,
            item_id=request.item_id,
            server_name=request.item_id,
            config_path="",
            error=str(e)
        )


@router.post("/uninstall", response_model=UninstallResponse)
async def uninstall_marketplace_item(
    request: UninstallRequest,
    user: User = Depends(get_current_user),
) -> UninstallResponse:
    """Uninstall an MCP server.

    Args:
        request: Uninstallation request with item ID.
        user: Current authenticated user.

    Returns:
        Uninstallation response with success status.
    """
    service = get_marketplace_service()

    item = service.get_item(request.item_id)
    if not item:
        # Item might have been removed from marketplace, but we can still uninstall
        # by creating a minimal item
        from makima.marketplace.models import MarketplaceItem
        item = MarketplaceItem(
            id=request.item_id,
            name=request.item_id,
            description="",
            content="{}",
        )

    try:
        response = service.uninstall_item(item=item, target=request.target)
        return response
    except Exception as e:
        logger.error("Failed to uninstall marketplace item", item_id=request.item_id, error=str(e))
        return UninstallResponse(
            success=False,
            item_id=request.item_id,
            server_name=request.item_id,
            config_path="",
            error=str(e)
        )


@router.get("/installed", response_model=list[InstalledItemInfo])
async def list_installed_items(
    target: str = Query("global", description="Installation target: 'project' or 'global'"),
    user: User = Depends(get_current_user),
) -> list[InstalledItemInfo]:
    """Get list of installed MCP servers.

    Args:
        target: Installation target to check.
        user: Current authenticated user.

    Returns:
        List of installed item information.
    """
    service = get_marketplace_service()
    return service.get_installed_items(target=target)


@router.get("/items/{item_id}/installed")
async def check_item_installed(
    item_id: str,
    target: str = Query("global", description="Installation target to check"),
    user: User = Depends(get_current_user),
) -> dict:
    """Check if a specific marketplace item is installed.

    Args:
        item_id: The item ID to check.
        target: Installation target to check.
        user: Current authenticated user.

    Returns:
        Dictionary with 'installed' boolean.
    """
    service = get_marketplace_service()
    is_installed = service.is_installed(item_id=item_id, target=target)
    return {"installed": is_installed, "item_id": item_id, "target": target}

File: apps/desktop-egui/src/api/marketplace.rs
use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// A parameter required for MCP server installation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpParameter {
    pub name: String,
    pub key: String,
    #[serde(default)]
    pub placeholder: String,
    #[serde(default)]
    pub optional: bool,
}

/// A specific installation method for an MCP server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpInstallationMethod {
    pub name: String,
    pub content: String,
    #[serde(default)]
    pub prerequisites: Vec<String>,
    #[serde(default)]
    pub parameters: Vec<McpParameter>,
}

/// An item in the MCP marketplace
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketplaceItem {
    pub id: String,
    pub name: String,
    pub description: String,
    pub author: Option<String>,
    pub author_url: Option<String>,
    pub url: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub prerequisites: Vec<String>,
    /// Either a string (single method) or array of methods
    pub content: MarketplaceContent,
    #[serde(default)]
    pub parameters: Vec<McpParameter>,
    /// Added by frontend - installation status
    #[serde(default)]
    pub installed: bool,
}

/// Marketplace content can be a single JSON string or an array of methods
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum MarketplaceContent {
    Single(String),
    Multiple(Vec<McpInstallationMethod>),
}

impl MarketplaceContent {
    /// Get the list of installation methods
    pub fn methods(&self) -> Vec<McpInstallationMethod> {
        match self {
            MarketplaceContent::Single(content) => {
                vec![McpInstallationMethod {
                    name: "Default".to_string(),
                    content: content.clone(),
                    prerequisites: vec![],
                    parameters: vec![],
                }]
            }
            MarketplaceContent::Multiple(methods) => methods.clone(),
        }
    }

    /// Check if there are multiple installation methods
    pub fn has_multiple_methods(&self) -> bool {
        matches!(self, MarketplaceContent::Multiple(methods) if methods.len() > 1)
    }
}

/// Response from listing marketplace items
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketplaceItemListResponse {
    pub items: Vec<MarketplaceItem>,
    pub total: usize,
}

/// Request to install an MCP server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstallRequest {
    pub item_id: String,
    pub target: String,
    pub selected_method_index: Option<usize>,
    pub parameters: HashMap<String, String>,
}

/// Response from installing an MCP server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstallResponse {
    pub success: bool,
    pub item_id: String,
    pub server_name: String,
    pub config_path: String,
    pub line: Option<usize>,
    pub error: Option<String>,
}

/// Request to uninstall an MCP server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UninstallRequest {
    pub item_id: String,
    pub target: String,
}

/// Response from uninstalling an MCP server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UninstallResponse {
    pub success: bool,
    pub item_id: String,
    pub server_name: String,
    pub config_path: String,
    pub error: Option<String>,
}

/// Information about an installed MCP server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstalledItemInfo {
    pub item_id: String,
    pub server_name: String,
    pub target: String,
    pub config_path: String,
    pub enabled: bool,
}

/// Response from checking if an item is installed
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstalledCheckResponse {
    pub installed: bool,
    pub item_id: String,
    pub target: String,
}

/// Marketplace API client
pub struct MarketplaceApi {
    client: Client,
    base_url: String,
    token: String,
}

impl MarketplaceApi {
    pub fn new(client: Client, base_url: String, token: String) -> Self {
        Self { client, base_url, token }
    }

    /// GET /api/marketplace/items - List all marketplace items
    pub async fn list_items(
        &self,
        search: Option<&str>,
        tags: Option<&[String]>,
    ) -> Result<MarketplaceItemListResponse> {
        let mut url = format!("{}/api/marketplace/items", self.base_url);
        let mut params = vec![];

        if let Some(s) = search {
            params.push(format!("search={}", urlencoding(s)));
        }
        if let Some(t) = tags {
            for tag in t {
                params.push(format!("tags={}", urlencoding(tag)));
            }
        }
        if !params.is_empty() {
            url = format!("{}?{}", url, params.join("&"));
        }

        let resp = self.client.get(&url).bearer_auth(&self.token).send().await
            .context("Failed to fetch marketplace items")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch marketplace items: {}", resp.status());
        }
        Ok(resp.json().await.context("Failed to parse marketplace items")?)
    }

    /// GET /api/marketplace/items/{id} - Get a single marketplace item
    pub async fn get_item(&self, item_id: &str) -> Result<MarketplaceItem> {
        let url = format!("{}/api/marketplace/items/{}", self.base_url, urlencoding(item_id));
        let resp = self.client.get(&url).bearer_auth(&self.token).send().await
            .context("Failed to fetch marketplace item")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch marketplace item: {}", resp.status());
        }
        Ok(resp.json().await.context("Failed to parse marketplace item")?)
    }

    /// GET /api/marketplace/tags - Get all available tags
    pub async fn list_tags(&self) -> Result<Vec<String>> {
        let url = format!("{}/api/marketplace/tags", self.base_url);
        let resp = self.client.get(&url).bearer_auth(&self.token).send().await
            .context("Failed to fetch marketplace tags")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch marketplace tags: {}", resp.status());
        }
        Ok(resp.json().await.context("Failed to parse marketplace tags")?)
    }

    /// POST /api/marketplace/install - Install an MCP server
    pub async fn install(&self, request: InstallRequest) -> Result<InstallResponse> {
        let url = format!("{}/api/marketplace/install", self.base_url);
        let resp = self.client.post(&url)
            .bearer_auth(&self.token)
            .json(&request)
            .send()
            .await
            .context("Failed to install marketplace item")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to install marketplace item: {}", resp.status());
        }
        Ok(resp.json().await.context("Failed to parse install response")?)
    }

    /// POST /api/marketplace/uninstall - Uninstall an MCP server
    pub async fn uninstall(&self, request: UninstallRequest) -> Result<UninstallResponse> {
        let url = format!("{}/api/marketplace/uninstall", self.base_url);
        let resp = self.client.post(&url)
            .bearer_auth(&self.token)
            .json(&request)
            .send()
            .await
            .context("Failed to uninstall marketplace item")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to uninstall marketplace item: {}", resp.status());
        }
        Ok(resp.json().await.context("Failed to parse uninstall response")?)
    }

    /// GET /api/marketplace/installed - Get list of installed items
    pub async fn list_installed(&self, target: &str) -> Result<Vec<InstalledItemInfo>> {
        let url = format!("{}/api/marketplace/installed?target={}", self.base_url, target);
        let resp = self.client.get(&url).bearer_auth(&self.token).send().await
            .context("Failed to fetch installed items")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch installed items: {}", resp.status());
        }
        Ok(resp.json().await.context("Failed to parse installed items")?)
    }

    /// GET /api/marketplace/items/{id}/installed - Check if an item is installed
    pub async fn is_installed(&self, item_id: &str, target: &str) -> Result<bool> {
        let url = format!(
            "{}/api/marketplace/items/{}/installed?target={}",
            self.base_url,
            urlencoding(item_id),
            target
        );
        let resp = self.client.get(&url).bearer_auth(&self.token).send().await
            .context("Failed to check installation status")?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to check installation status: {}", resp.status());
        }
        let check: InstalledCheckResponse = resp.json().await
            .context("Failed to parse installation check")?;
        Ok(check.installed)
    }
}

fn urlencoding(s: &str) -> String {
    url::form_urlencoded::byte_serialize(s.as_bytes()).collect()
}